### PR TITLE
Custom Basis Class

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
## Add `CustomBasis` and Refactor Core Basis Infrastructure

This PR introduces a new `CustomBasis` class, which allows users to define a fully functional NeMoS basis by providing a list of callable functions. Each callable acts as an individual basis function.

The `CustomBasis` is designed to behave as closely as possible to a standard `nemos` basis. In particular:

- `compute_features` automatically vectorizes over extra input dimensions;
- Label generation and basis composition work seamlessly with other NeMoS components.

Vectorization is implemented using `itertools` and array slicing. While a JAX `vmap` would have been more efficient, we deliberately avoid assuming that user-defined functions are JAX-compatible to maintain generality.

The main difference in the public API is that `CustomBasis` does **not** implement `evaluate_on_grid`, since this may not make sense in all use cases — for example, when the basis processes images or other high-dimensional inputs.

---

### 🔧 New `BasisMixin` Class

A new `BasisMixin` class consolidates logic shared across all basis classes. Common functionality previously implemented in `nemos.basis._basis.Basis` has been moved to this mixin.

The following methods are now implemented in `BasisMixin`:

- `split_by_feature`
- `_generate_subtree_labels`
- `_iterate_over_components`
- `label` (getter and setter)
- `set_input_shape`
- `input_shape` (getter)
- `_get_feature_slicing`
- `_reshape_concatenated_arrays`
- `__iter__`

---

### ✅ Basis Input Validation

Validation logic for basis inputs has been moved to a new module: `_check_basis.py`. This modularization improves code clarity and separation of concerns.

---

### 🧩 Composition Utilities

A new set of utility functions handle composition logic for basis objects, reducing reliance on private methods.

These utilities cover:

- Label generation for composite bases
- Label setter behavior
- Input shape handling (`set_input_shape`, `get_input_shape`)
- API compliance checks for basis-like objects
- Arithmetic operations on bases (multiplication and exponentiation by positive integers)
- Promotion of outputs to transformer bases

### CODE REVIEW

**Review by sjvenditto — CHANGES_REQUESTED** [review]
Great PR! The custom basis is very well done! I noted a few places with minor comments/questions

File: src/nemos/basis/_custom_basis.py [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+        self._pynapple_support = bool(pynapple_support)
+        self.funcs = funcs
+        self.ndim_input = int(ndim_input)
+
+        if output_shape is None:
+            output_shape = ()
+
+        self.output_shape = output_shape
+
+        # nomenclature is confusing, should rename this to _n_args_compute_features
```
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/nemos/basis/_custom_basis.py:206]: I'll do the change!

[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/nemos/basis/_custom_basis.py:206]: after a discussion, we can keep it as a private attribute!

File: docs/how_to_guide/define_a_custom_basis.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+# partial() will bind 'poly_coef' and 'decay_rate'
+# as keyword arguments, leaving 'x' as a keyword-only parameter.
+vmap_laguerre = jax.vmap(laguerre_poly, in_axes=0, out_axes=1)
+f = partial(vmap_laguerre, poly_coef=P, decay_rate=c)
+print(inspect.signature(f))
+
+# Calling f(x) positionally confuses vmap’s shape inference (it expects three positional args),
+# so it fails with a shape/axis error before reaching laguerre_poly.
+f(x)
```
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply docs/how_to_guide/define_a_custom_basis.md:116]: I don't know how to have the cell executed when it is nested in an admonition. Do you know how to do that? otherwise what do you suggest?

File: docs/how_to_guide/define_a_custom_basis.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+# define masks using a nemos 2D basis
+basis_2d = nmo.basis.RaisedCosineLinearEval(8)**2
+_, _, masks = basis_2d.evaluate_on_grid(50, 50)
+funcs = [partial(image_dot_product, mask=m) for m in masks.T]
+
+# specify the the expected 3D input, (n_samples, pixel, pixel)
+bas_img = nmo.basis.CustomBasis(funcs=funcs, ndim_input=3, label="Image-dot")
+
+print(bas_img.compute_features(imgs).shape)
```
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply docs/how_to_guide/define_a_custom_basis.md]: I tried to create a better example, let me know!

[the remark this answers is not in the document]
Comment (sjvenditto) [reply docs/how_to_guide/define_a_custom_basis.md]: this new one looks great!

File: docs/how_to_guide/define_a_custom_basis.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjvenditto) [reply docs/how_to_guide/define_a_custom_basis.md:117]: I don't know a good way other than hard coding the error output, but I don't think it's necessary

**Comment by BalzaniEdoardo** [comment]
Thanks @sjvenditto for the review!
